Add Snowflake support with standardized URL parsing

ctdunc opened the pull request.

ctdunc committed:
initial adapter commit. currently only supports connecting using snowsql config file

ctdunc committed:
got snowflake working with reasonable url defaults

ctdunc committed:
add snowflake connection doc

ctdunc committed:
add snowflake to readme

ctdunc committed:
Merge branch 'tpope:master' into snowflake

ctdunc committed:
Remove unnecessary comments in snowflake.vim

ctdunc commented:
Wasn't sure how to update #65 with my own changes, so opened a new PR here. Please let me know if any changes required!

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,44 @@
+
+function! s:command_for_url(url) abort
+  let url = db#url#parse(a:url)
+  "extra options here turn off nonsense before/after results
```
I don't think we want this for `#interactive()`. I think what you want is to provide a separate `#filter()` function, and add the output modifying options there.

ctdunc replied:
I don't have enough experience with DB to know whether this is true. The friendly/timing options show a hello/goodbye message that would show up in interactive mode before and after the actual returned data. Query timing is already handled by dadbod, so I think it might make sense to suppress these. 

Have made the requested change for now, but let me know your thoughts.

tpope replied:
`#interactive()` is the command used when you call `:DB snowflake://` with no query to start an interactive session. This doesn't have any sort of timing.

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,44 @@
+endfunction
+
+function! db#adapter#snowflake#interactive(url) abort
+  echomsg "starting snowsql. executing query"
```
Looks like a debug statement that should be removed.

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,44 @@
+  return s:command_for_url(a:url)
+endfunction
+
+function! db#adapter#snowflake#input_flag() abort
+  return ' -f '
+endfunction
```
`#input_flag()` was removed a while back. You want something like this instead:

```vim
function! db#adapter#snowflake#input(url, in) abort
  return db#adapter#snowflake#filter(a:url) + ['-f', a:in]
endfunction
```

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,44 @@
+  let cmd = s:command_for_url(a:url)
+  let cmd .= ' --query "show databases"'
+  " if cmd looks to have no cli args, treat it as a string
+  let out = system(cmd)
```
Needs to be retooled to use `db#systemlist()`.

ctdunc replied:
Retooled this, however have to apply a map after calling `systemlist` since snowflake doesn't provide a way to get database names _only_ in the same way e.g. MariaDB does without having an active warehouse.

ctdunc committed:
Fix snowflake DB completion, use of outdated methods

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,45 @@
+if exists('g:autoloaded_db_snowflake')
+  finish
+endif
+let g:autoloaded_db_snowflake = 1
```
No longer necessary.

tpope commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,45 @@
+function! db#adapter#snowflake#complete_database(url) abort
+  let pre = matchstr(a:url, '[^:]\+://.\{-\}/')
+  let cmd = s:command_for_url(pre) +
+      \['-q', 'show terse databases'] + 
+      \s:no_timing_friendly + 
+      \['-o', 'header=false', '-o', 'output_format=tsv'] 
```
Line continuations should be formatted as 6 spaces, a backslash, and one more space. Same elsewhere in the file.

ctdunc committed:
Resolve interactive filter issue, formatting

pprotas commented on autoload/db/adapter/snowflake.vim:
```diff
@@ -0,0 +1,45 @@
+  let url = db#url#parse(a:url)
+  "extra options here turn off nonsense before/after results
+  let cmd = (has_key(url, 'password') ? ['env', 'SNOWSQL_PWD=' . url.password] : []) +
+      \ [ "snowsql" ] +
```
Hi, hope y'all don't mind if I chime in here a bit. I've been testing your branch to query Snowflake from `vim-dadbod` and `vim-dadbod-ui`, and here is some observations I made during usage:

## SSO

My company uses Okta to auth with Snowflake. This means I need to use the `--authenticator=externalbrowser` query param. When using `snowsql`, this means that a browser window pops up, and asks me to verify with Okta every time I want to perform a query. This is very frustrating.

I found that Snowflake is creating the `snow` CLI right now: https://docs.snowflake.com/en/developer-guide/snowflake-cli-v2/index

This tool has an automatic caching of the Okta credentials. So I only have to auth once, and every query I send afterwards does not force me to authorize again. Maybe using this newer tool would prove beneficial for SSO usecases? Let me know what you think. The tool is still experimental, but seems to work very well. Let me know if I can contribute.

## Email as username

This is a smal nit, but the `snowflake://username@accountname` querystring syntax doesn't work if your username is an email, just like in my case. I have to use `snowflake://accountname?username=<email>` instead. Not a big issue, just wanted to mention this.

ctdunc replied:
`snow` seems quite nice -- hadn't seen this yet. I can take a look at incorporating next week, or would accept a pr into my branch. zsh completion and credentials are two very nice QOL improvements.

Not sure how to solve the email as username issue given the syntax requirements, but it seems like that workaround is fine. Could maybe re-add support for the `snowflake://connection` syntax for folks who want to use global or workspace config file.

tpope replied:
> This is a smal nit, but the `snowflake://username@accountname` querystring syntax doesn't work if your username is an email, just like in my case.

https://github.com/tpope/vim-dadbod/blob/7888cb7164d69783d3dce4e0283decd26b82538b/doc/dadbod.txt#L47-L51

orest-tokovenko-deel replied:
Looking forward to using this with Snowflake support!

daasity-nathan commented:
When I use this, it seems like any query I write hangs out for a few (7-10) seconds, before starting the "Execute query" loading screen (initiating the connection via snowsql?). On the other hand, using DB to connect to BigQuery seems to start the "Execute query" screen much faster (0-2 seconds). 

Is there some way you are using to step around this that I'm not seeing @ctdunc?

ctdunc commented:
> When I use this, it seems like any query I write hangs out for a few (7-10) seconds, before starting the "Execute query" loading screen (initiating the connection via snowsql?). On the other hand, using DB to connect to BigQuery seems to start the "Execute query" screen much faster (0-2 seconds). 
> 
> Is there some way you are using to step around this that I'm not seeing @ctdunc?

I see a similar time for connection (not quite as long, does it take u this long to connect via the CLI normally?). My initial thought is that this is probably a limitation of the snowsql tool. 

I think any kind of keep-alive is probably beyond the scope of dadbod, but i can look a bit further. If you find a workaround for this please let me know!

ctdunc commented:
@tpope is there anything else blocking this PR? I think I've addressed your comments, happy to make any more suggested changes. 

Ill hopefully have time to dedicate to looking at snowflakes new CLI tools for that one SSO caching issue, but i think as-is we have 95% of use cases covered, and it doesnt seem like this functionality is included with any of the other databases. 

Thanks again for your time and patience!

vanducng commented:
@ctdunc  It works fairly great for me. Just curious if there is additional config to display all the databases?

![CleanShot 2024-08-11 at 23 46 07@2x](https://github.com/user-attachments/assets/94c2c29f-73e8-4911-aa29-ccd6bca5ae05)

ctdunc commented:
> @ctdunc  It works fairly great for me. Just curious if there is additional config to display all the databases?
> 
> ![CleanShot 2024-08-11 at 23 46 07@2x](https://github.com/user-attachments/assets/94c2c29f-73e8-4911-aa29-ccd6bca5ae05)
> 

Looks like you are using DBUI, which is a different project. https://github.com/kristijanhusak/vim-dadbod-ui

I have spent maybe 20 mins on getting this working in DBUI & have a rough version working, but needs cleanup. probably need approval here first before submitting a pr there as this is a dependency

vanducng commented:
> > @ctdunc  It works fairly great for me. Just curious if there is additional config to display all the databases?
> > ![CleanShot 2024-08-11 at 23 46 07@2x](https://private-user-images.githubusercontent.com/2980905/356892794-94c2c29f-73e8-4911-aa29-ccd6bca5ae05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MTM5MDcsIm5iZiI6MTcyMzQxMzYwNywicGF0aCI6Ii8yOTgwOTA1LzM1Njg5Mjc5NC05NGMyYzI5Zi03M2U4LTQ5MTEtYWEyOS1jY2Q2YmNhNWFlMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMjIwMDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDA4ZjJiMTdkNGFkMTZkZjEzODhkMGE1ZGZmMWNjZTViMmU0ZmY3NTc5NDJlZmIyZmQ3NWE5ODUzNTA0ZmNkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gCfZnjRvRPR61-iGGI496_zi4L2ZkbrbJ9-mSoSemKQ)
> 
> Looks like you are using DBUI, which is a different project. https://github.com/kristijanhusak/vim-dadbod-ui
> 
> I have spent maybe 20 mins on getting this working in DBUI & have a rough version working, but needs cleanup. probably need approval here first before submitting a pr there as this is a dependency.

Awesome, Im very eager to try it out for my VSCode Snowflake extension replacement.

ctdunc commented:
> > > @ctdunc  It works fairly great for me. Just curious if there is additional config to display all the databases?
> > > ![CleanShot 2024-08-11 at 23 46 07@2x](https://private-user-images.githubusercontent.com/2980905/356892794-94c2c29f-73e8-4911-aa29-ccd6bca5ae05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MTM5MDcsIm5iZiI6MTcyMzQxMzYwNywicGF0aCI6Ii8yOTgwOTA1LzM1Njg5Mjc5NC05NGMyYzI5Zi03M2U4LTQ5MTEtYWEyOS1jY2Q2YmNhNWFlMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMjIwMDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDA4ZjJiMTdkNGFkMTZkZjEzODhkMGE1ZGZmMWNjZTViMmU0ZmY3NTc5NDJlZmIyZmQ3NWE5ODUzNTA0ZmNkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gCfZnjRvRPR61-iGGI496_zi4L2ZkbrbJ9-mSoSemKQ)
> > 
> > Looks like you are using DBUI, which is a different project. https://github.com/kristijanhusak/vim-dadbod-ui
> > 
> > I have spent maybe 20 mins on getting this working in DBUI & have a rough version working, but needs cleanup. probably need approval here first before submitting a pr there as this is a dependency.
> 
> Awesome, Can you share your fork on `vim-dadbod-ui` ? Im very eager to try it out for my VSCode Snowflake extension replacement. 

Sure. github.com/ctdunc/vim-dadbod-ui-snowflake 

Still really rudimentary, no autocomplete support etc. Going on vacation shortly but happy to try out any features once back in internet land

vanducng commented:
> > > > @ctdunc  It works fairly great for me. Just curious if there is additional config to display all the databases?
> > > > ![CleanShot 2024-08-11 at 23 46 07@2x](https://private-user-images.githubusercontent.com/2980905/356892794-94c2c29f-73e8-4911-aa29-ccd6bca5ae05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0MTM5MDcsIm5iZiI6MTcyMzQxMzYwNywicGF0aCI6Ii8yOTgwOTA1LzM1Njg5Mjc5NC05NGMyYzI5Zi03M2U4LTQ5MTEtYWEyOS1jY2Q2YmNhNWFlMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTFUMjIwMDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDA4ZjJiMTdkNGFkMTZkZjEzODhkMGE1ZGZmMWNjZTViMmU0ZmY3NTc5NDJlZmIyZmQ3NWE5ODUzNTA0ZmNkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gCfZnjRvRPR61-iGGI496_zi4L2ZkbrbJ9-mSoSemKQ)
> > > 
> > > 
> > > Looks like you are using DBUI, which is a different project. https://github.com/kristijanhusak/vim-dadbod-ui
> > > I have spent maybe 20 mins on getting this working in DBUI & have a rough version working, but needs cleanup. probably need approval here first before submitting a pr there as this is a dependency.
> > 
> > 
> > Awesome, Can you share your fork on `vim-dadbod-ui` ? Im very eager to try it out for my VSCode Snowflake extension replacement.
> 
> Sure. github.com/ctdunc/vim-dadbod-ui-snowflake
> 
> Still really rudimentary, no autocomplete support etc. Going on vacation shortly but happy to try out any features once back in internet land

It is a great start, I start trying out now, thank you for putting effort on it. 

For this PR, could we make it main if no additional updates required @tpope ?

ctdunc committed:
Merge branch 'tpope:master' into snowflake

Slanman3755 commented:
+1 on getting this moving

tpope committed:
Clean up Snowflake documentation

tpope committed:
Clean up Snowflake implementation

tpope commented:
I did a pass of cleanup. Please sanity check that everything's still working as intended. Once you confirm, I'll squash and merge.

vanducng commented:
hey @ctdunc, just one more step. Im exited to switch to main branch now.

ctdunc commented:
LGTM--connecting works and can execute queries! Thank you!
